src/Features/CSharp/Portable/MakeStructFieldsWritable/CSharpMakeStructFieldsWritableDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,99 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.CSharp.MakeStructFieldsWritable
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpMakeStructFieldsWritableDiagnosticAnalyzer : AbstractCodeQualityDiagnosticAnalyzer
+    {
+        private static readonly DiagnosticDescriptor s_diagnosticDescriptor = CreateDescriptor(
+            IDEDiagnosticIds.MakeStructFieldsWritable,
+            new LocalizableResourceString(nameof(FeaturesResources.Make_readonly_fields_writable), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            new LocalizableResourceString(nameof(FeaturesResources.Make_readonly_fields_writable), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: false);
+
+        public CSharpMakeStructFieldsWritableDiagnosticAnalyzer()
+            : base(ImmutableArray.Create(s_diagnosticDescriptor), GeneratedCodeAnalysisFlags.ReportDiagnostics)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        {
+            return DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => true;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext
+                => CompilationAnalyzer.CreateAndRegisterActions(compilationStartContext));
+        }
+
+        private sealed class CompilationAnalyzer
+        {
+            private bool _hasReadonlyField = false;
+            private bool _hasTypeInstanceAssigment = false;
+            private SyntaxNode _typeInstanceAssigmentNode;
+
+            public static void CreateAndRegisterActions(CompilationStartAnalysisContext compilationStartContext)
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    // We are only interested in struct declarations
+                    var namedTypeSymbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+                    if (namedTypeSymbol.TypeKind != TypeKind.Struct) return;
+
+                    var compilationAnalyzer = new CompilationAnalyzer();
+                    compilationAnalyzer.RegisterActions(symbolStartContext);
+                }, SymbolKind.NamedType);
+            }
+
+            private void RegisterActions(SymbolStartAnalysisContext symbolStartContext)
+            {
+                // We report diagnostic only if these requirements are met:
+                // 1. The type is struct
+                // 2. Struct contains at least one 'readonly' field
+                // 3. Struct contains assignment to 'this' outside the scope of constructor
+
+                symbolStartContext.RegisterSyntaxNodeAction(AnalyzeIfFieldIsReadonly, SyntaxKind.FieldDeclaration);
```

Comment:
>RegisterSyntaxNodeAction [](start = 35, length = 24)

You don't need to register an action for this - you can directly cast `symbolStartContext.Symbol` to `INamedTypeSymbol` and then invoke `GetMembers().OfType<IFieldSymbol>()` to get all the fields.

Reply:
Also it seems like this was the only reason why you had to make this a language specific analyzer in the CSharp Features layer. Once you remove the above `RegisterSyntaxNodeAction`, you can move this analyzer into the core layer so it has a language agnostic implementation.

---
In reply to: [244890718](https://github.com/dotnet/roslyn/pull/32063#discussion_r244890718) [](ancestors = 244890718)

Reply:
> I understand that feature should be supported only for C#.

Thanks. I would still recommend not registering an unnecessary syntax node action to just identify field symbols - you can already access these symbols directly: `((INamedTypeSymbol)symbolStartContext.Symbol).GetMembers().OfType<IFieldSymbol>())`

Reply:
So, I've tried to use `((INamedTypeSymbol)symbolStartContext.Symbol).GetMembers().OfType<IFieldSymbol>())` and I realize it changes behaviour slightly.

The analyzer reports struct declaration with readonly properties `public int Value { get; }` as readonly fields. But I think this is more correct and precise behavior for analyzer, so I will keep this way.

@mavasani 
thanks for suggestion.

Reply:
I've been thinking about this lately and I came to the conclusion, that it won't be trivial to create proper implementation of writable properties, as it was in case of `readonly`  fields.
It's easy to create implementation for 
`public int Value { get; }` ---> `public int Value { get; set; }` , but it starts to be more complex for case like 
`public int Value { get { return _value; } }` ---> `public int Value { get { return _value; } set { _value = value; } }`.
I looked at `edit.Generator` and there are some methods which could be usefull but I'm not sure if it is enough.

Therefor I would like to hear your opinion @mavasani and I would be grateful if you could point me to the API which could help me, or Analyzer which solves similar problem.

Thanks

Reply:
@KristianJakubik I would recommend not dealing with properties for the initial implementation, and only consider readonly fields, as mentioned [here](https://github.com/dotnet/roslyn/issues/30263#issue-365888322). So, even for readonly properties such as `public int Value { get; }`, it is fine to not have any diagnostic. You can filter out implicitly declared backing fields as `((INamedTypeSymbol)symbolStartContext.Symbol).GetMembers().OfType<IFieldSymbol>().Where(f => f.AssociatedSymbol == null)`

Reply:
Ok, then lets keep only for fields.

Thanks

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/32063/commits/6a5558daa8a7f20616c80b989b8886fbe3cbd2cf](https://github.com/dotnet/roslyn/pull/32063/commits/6a5558daa8a7f20616c80b989b8886fbe3cbd2cf)